src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc

```diff
@@ -607,14 +658,81 @@ void PickFirst::SubchannelList::SubchannelData::OnConnectivityStateChange(
   // see its initial notification.  Start trying to connect, starting
   // with the first subchannel.
   if (!old_state.has_value()) {
-    subchannel_list_->subchannels_.front().ReactToConnectivityStateLocked();
+    if (!IsPickFirstHappyEyeballsEnabled()) {
+      subchannel_list_->subchannels_.front().ReactToConnectivityStateLocked();
+      return;
+    }
+    subchannel_list_->StartConnectingNextSubchannel();
+    return;
+  }
+  if (!IsPickFirstHappyEyeballsEnabled()) {
+    // Ignore any other updates for subchannels we're not currently trying to
+    // connect to.
+    if (Index() != subchannel_list_->attempting_index_) return;
+    // React to the connectivity state.
+    ReactToConnectivityStateLocked();
     return;
   }
-  // Ignore any other updates for subchannels we're not currently trying to
-  // connect to.
-  if (Index() != subchannel_list_->attempting_index_) return;
-  // React to the connectivity state.
-  ReactToConnectivityStateLocked();
+  // Otherwise, process connectivity state change.
```

Comment:
nit: I feel like SubchannelData::OnConnectivityStateChange is becoming really long and hard to follow. Should it be split in several methods?

Reply:
I think it will be easier to read once the experiment goes away and we have only the new code to deal with.